synapse/storage/databases/main/account_data.py

```diff
@@ -64,14 +63,12 @@ def __init__(
     ):
         super().__init__(database, db_conn, hs)
 
-        # `_can_write_to_account_data` indicates whether the current worker is allowed
-        # to write account data. A value of `True` implies that `_account_data_id_gen`
-        # is an `AbstractStreamIdGenerator` and not just a tracker.
-        self._account_data_id_gen: AbstractStreamIdTracker
```

Comment:
I guess this is no longer true and we're always a writer?

Reply:
Mm, no, not all workers are account_data stream writers. But `self._account_data_id_gen` *is* always a `AbstractStreamIdGenerator`.

The methods that rely on `self._account_data_id_gen` being a `AbstractStreamIdGenerator` should not be called unless we are a writer though. Hence the `assert`s in those methods, such as:

https://github.com/matrix-org/synapse/blob/c8665dd25d18fa7d7176984cee191834002909a0/synapse/storage/databases/main/account_data.py#L772

Defining `self._account_data_id_gen` as a `AbstractStreamIdTracker` confused mypy though, as `AbstractStreamIdTracker` doesn't have the methods that a writer need (`get_next`, `get_next_txn`, etc.).

mypy didn't yell about `get_next` not existing on `AbstractStreamIdTracker` though, possibly because we used it in `async with ...` blocks instead of calling it directly? Does mypy check those properly?

Reply:
It looks like #14468 essentially removed `AbstractStreamIdTracker`'s concrete classes without removing the abstract class, which seems unfortunate.

Reply:
It seems there's some confusion about making sequence generators `AbstractStreamIdTracker` in `*WorkerStore` classes, where it's assumed that the same sequence generator would be an `AbstractStreamIdGenerator` in the non-worker variant storage class. See `ReceiptsWorkerStore` for instance:

https://github.com/matrix-org/synapse/blob/9bb2eac71962970d02842bca441f4bcdbbf93a11/synapse/storage/databases/main/receipts.py#L66-L68

But `*WorkerStore` doesn't mean that that storage class can't write to the database. It just means that workers should use that class vs. the main process... right?

Reply:
> It seems there's some confusion about making sequence generators `AbstractStreamIdTracker` in `*WorkerStore` classes, where it's assumed that the same sequence generator would be an `AbstractStreamIdGenerator` in the non-worker variant storage class. See `ReceiptsWorkerStore` for instance:

#14468 made this obsolete as far as I can tell. So they just never got updated. 🤷

Reply:
See #15192.